### DISCUSSION TITLE
A little help migrating from Ethers

### DISCUSSION THREAD

**Discussion by rgarcia1129** [wevm/viem#627] [Question]
Hi all, 
I'm working on the server side looking to rewrite the following function using viem instead of ethers. It seems simple enough and I've kinda made some progress but the hashes continue to not match and the sign message gives me an error of "personal_sign" is not supported.

Any help would be appreciated thanks! 

  const provider = new ethers.providers.AlchemyProvider(
    "homestead",
    "<ALCHEMY_API_KEY"
  );

  const signer = new Wallet(
    "<SIGNER_ADDRESS>",
    provider
  );

  const hash = ethers.utils.solidityKeccak256(
    ["bytes"],
    [
      ethers.utils.defaultAbiCoder.encode(
        ["address", "uint256"],
        [minter, salt]
      ),
    ]
  );
  const signature = await signer.signMessage(ethers.utils.arrayify(hash));

return signature

**Comment by rgarcia1129** [comment]
I know that signer would be defined like this so Im really just looking for matching the hashes and generating the signature. 

 ` const signer = createWalletClient({
    account: privateKeyToAccount(
      "0x" + <PRIVATE_KEY>"
    ),
    chain: mainnet,
    transport: http(
      "https://eth-mainnet.g.alchemy.com/v2/<ALCHEMY_KEY"
    )
  });`

**Reply by jxom** [reply]
You may be getting signature differences because viem does not support signing of raw hex/bytes data yet. There is actually a [PR open for that right now](https://github.com/wagmi-dev/viem/pull/610), and it should be released by the end of the week.

Here is an viem equivalent example of the code you gave:

```ts
import { createWalletClient, http, keccak256, encodeAbiParameters, encodePacked } from 'viem'
import { privateKeyToAccount } from 'viem/accounts' 

const walletClient = createWalletClient({
  account: privateKeyToAccount('0x...')
  chain: mainnet,
  transport: http('https://eth-mainnet.alchemyapi.io/v2/ALCHEMY_API_KEY')
})

const hash = keccak256(
  encodePacked(
    ['bytes'], 
    [encodeAbiParameters(['address', 'uint256'], [minter, salt])]
  )
 )

await walletClient.signMessage({ message: { raw: hash } })
```

Note: the `raw` attribute in `signMessage` won't work yet! Waiting until #610 is merged.

**Reply by rgarcia1129** [reply]
thanks for the quick response! 
when using this code snippet to get the hash I get the following error: `Argument of type '(bigint | `0x${string}`)[]' is not assignable to parameter of type 'never'.ts(2345) `

<img width="491" alt="image" src="https://github.com/wagmi-dev/viem/assets/134097400/c5536ac6-09c8-4970-bee5-a15a7d7ff23d">

**Reply by wslyvh** [reply]
I realize this an old discussion item, but ended up here while searching to migrate from `defaultAbiCoder.encode`
So sharing here for others..

The format for encodeAbiParameters is incorrect and you need to specify the types like
```
const encodedData = encodeAbiParameters(
    [{ type: 'address' }, { type: 'uint256' }],
    [minter, salt]
)
```

**Comment by rgarcia1129** [comment]
I figured out that these are generating equivalent hashes. Do you see any issues with this viem method of generating a hash?

```
  const encodedParams = ethers.utils.defaultAbiCoder.encode(
    ["address", "uint256"],
    [minter, salt]
  );

  const hash = ethers.utils.solidityKeccak256(["bytes"], [encodedParams]);
```


```
  const encodedParams = pad(
    encodePacked(["address", "uint256"], [minter, salt]),
    {
      dir: "left",
      size: 64,
    }
  );

  const hash = keccak256(encodedParams);
```

**Reply by jxom** [reply]
Should be fine!

**Comment by rgarcia1129** [comment]
will keep an eye that pr for the message signature 👍

**Reply by jxom** [reply]
Released!

**Reply by rgarcia1129** [reply]
already updated, it worked like a charm 👍  thanks for getting this out so quick

**Comment by sbauch** [comment]
I had some trouble getting this figured out, so in case anyone hits a similar issue and comes across this, this is the key aspect that helped me migrate from ethers:

> You may be getting signature differences because viem does not support signing of raw hex/bytes data yet. There is actually a https://github.com/wagmi-dev/viem/pull/610, and it should be released by the end of the week.

`signMessage` by default adds the [EIP-191 format](https://eips.ethereum.org/EIPS/eip-191) to the front of the signature. So you probably want to use this:

```
  const signature = await account.signMessage({ message: { raw: message } });
```
